colossalai/shardformer/modeling/falcon.py

```diff
@@ -0,0 +1,774 @@
+
+        query_layer_ = query_layer.reshape(batch_size, self.num_heads, -1, self.head_dim).transpose(1, 2).contiguous()
+        key_layer_ = key_layer.reshape(batch_size, num_kv_heads, -1, self.head_dim).transpose(1, 2).contiguous()
+        value_layer_ = value_layer.reshape(batch_size, num_kv_heads, -1, self.head_dim).transpose(1, 2).contiguous()
+
+        if alibi is not None:
+            attention_mask_float = (
+                attention_mask_float + alibi.view(batch_size, self.num_heads, 1, kv_length) * self.beta
+            )
+ 
+        batch_size, src_len = query_layer_.size()[0], query_layer_.size()[1]
+        tgt_len = key_layer_.size()[1]
+        attention_mask_float = attention_mask_float.expand(
+            batch_size, self.num_heads, src_len, tgt_len
+        ).contiguous()
+        context_layer = me_attention(
+            query_layer_,
+            key_layer_,
+            value_layer_,
+            attn_bias=attention_mask_float,
+            scale=self.inv_norm_factor,
+            p=self.attention_dropout.p,
+        )
+        batch_size, seq_length, _, _ = context_layer.shape
+        context_layer = context_layer.reshape(batch_size, seq_length, -1)
+        
+        output_tensor = self.dense(context_layer)
+
+        return output_tensor, present
+
+    return forward
+
+
+class FalconPipelineForwards:
+    """
+    This class serves as a micro library for falcon pipeline forwards.
+    """
+
+    @staticmethod
+    def falcon_model_forward(
+        self: FalconModel,
+        input_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.Tensor, torch.Tensor], ...]] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        head_mask: Optional[torch.LongTensor] = None,
+        inputs_embeds: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        stage_manager: Optional[PipelineStageManager] = None,
+        hidden_states: Optional[torch.FloatTensor] = None,
+        stage_index: Optional[List[int]] = None,
+        shard_config: ShardConfig = None,
+    ) -> Union[Tuple[torch.Tensor, ...], BaseModelOutputWithPastAndCrossAttentions]:
+        logger = logging.get_logger(__name__)
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if past_key_values is None:
+            past_key_values = tuple([None] * len(self.h))
```

Comment:
past_key_values is not supported for Shardformer pipeline, please warn

Reply:
updated